src/DbPostgreSQLAdapter/Database.cpp

```diff
@@ -0,0 +1,134 @@
+#include "stdafx.h"
+
+#include "Database.h"
+#include "DbAdapterInterface/ITable.h"
+#include "RecordSet.h"
+#include "Table.h"
+#include "TableRecordSet.h"
+#include "Transaction.h"
+
+#include <libpq-fe.h>
+
+namespace {
+	std::unique_ptr<PGresult, void(*)(PGresult*)> createRAIIPGresult(PGresult* result)
+	{
+		return std::unique_ptr<PGresult, void(*)(PGresult*)>(result, PQclear);
+	}
+}
+
+namespace systelab::db::postgresql {
+
+	Database::Database(PGconn* database)
+		: m_database(database)
+		, m_lastOperationRowsAffected(0)
+		, m_lastInsertedRowId(0)
+	{
+	}
+
+	Database::~Database()
+	{
+		PQfinish(m_database);
+	}
+
+	ITable& Database::getTable(std::string tableName)
+	{
+		const auto itr = m_tables.find(tableName);
+		if (itr == m_tables.cend())
+		{
+			m_tables.insert(std::make_pair(tableName, new Table(*this, tableName)));
+		}
+
+		const auto& table = m_tables.at(tableName);
+		return *(table.get());
+	}
+
+	std::unique_ptr<IRecordSet> Database::executeQuery(const std::string& query)
+	{
+		std::lock_guard<std::recursive_mutex> lock(m_mutex);
+		const auto statementResult = createRAIIPGresult(PQexec(m_database, query.data()));
+		if (PQresultStatus(statementResult.get()) == PGRES_TUPLES_OK)
+		{
+			return std::make_unique<RecordSet>(statementResult.get());
```

Comment:
Actually I think that it would be better to change the constructors to reflect better the ownership. 

From my POV, RecordSet does not have to take ownership of this object as it's only purpose is to use it in order to get initialized. The true owner of the object would be database who, in turn should be responsible of deleting it when the scope finishes. 

 What I can do is to change the constructors to accept a const unique_ptr<PGResult>&

Reply:
Ok, I went trough a lot of tries to see what was the best approach for this specific case, and it seems that the current approach is the best one (if we can say so) for the following reasons:

* _release_ method is not supported when a unique_ptr with custom destructor is being used.
* Changing the signature of the different methods (as I suggested first) seems a good approach, but then we have to pass a const reference to a special unique_ptr type (again, due to the destructor), so we would end with a lot of typedefs (or horrible method signatures) around.